evm/src/cpu/kernel/asm/bignum/add.asm

```diff
@@ -0,0 +1,87 @@
+// Arithmetic on little-endian integers represented with 128-bit limbs.
+// All integers must be under a given length bound, and are padded with leading zeroes.
+
+// Replaces a with a + b, leaving b unchanged.
+global add_bignum:
+    // stack: len, a_start_loc, b_start_loc, retdest
+    %stack (l, a, b) -> (0, 0, a, b, l)
+    // stack: carry=0, i=0, a_start_loc, b_start_loc, len, retdest
+add_loop:
+    // stack: carry, i, a_i_loc, b_i_loc, len, retdest
+    DUP4
+    %mload_kernel_general
+    // stack: b[i], carry, i, a_i_loc, b_i_loc, len, retdest
+    DUP4
+    %mload_kernel_general
+    // stack: a[i], b[i], carry, i, a_i_loc, b_i_loc, len, retdest
+    ADD
+    ADD
+    // stack: a[i] + b[i] + carry, i, a_i_loc, b_i_loc, len, retdest
+    %stack (val) -> (val, @BIGNUM_LIMB_BASE, @BIGNUM_LIMB_BASE, val)
+    // stack: a[i] + b[i] + carry, 2^128, 2^128, a[i] + b[i] + carry, i, a_i_loc, b_i_loc, len, retdest
+    DIV
+    // stack: (a[i] + b[i] + carry) // 2^128, 2^128, a[i] + b[i] + carry, i, a_i_loc, b_i_loc, len, retdest
+    SWAP2
+    // stack: a[i] + b[i] + carry, 2^128, (a[i] + b[i] + carry) // 2^128, i, a_i_loc, b_i_loc, len, retdest
+    MOD
+    // stack: c[i] = (a[i] + b[i] + carry) % 2^128, carry_new = (a[i] + b[i] + carry) // 2^128, i, a_i_loc, b_i_loc, len, retdest
+    DUP4
+    // stack: a_i_loc, c[i], carry_new, i, a_i_loc, b_i_loc, len, retdest
+    %mstore_kernel_general
+    // stack: carry_new, i, a_i_loc, b_i_loc, len, retdest
+    %stack (c, i, a, b) -> (a, b, c, i)
+    // stack: a_i_loc, b_i_loc, carry_new, i, len, retdest
+    %increment
+    SWAP1
+    %increment
+    SWAP1
+    %stack (a, b, c, i) -> (c, i, a, b)
+    // stack: carry_new, i, a_i_loc + 1, b_i_loc + 1, len, retdest
+    SWAP1
+    %increment
+    SWAP1
+    // stack: carry_new, i + 1, a_i_loc + 1, b_i_loc + 1, len, retdest
+    DUP5
+    DUP3
+    // stack: i + 1, len, carry_new, i + 1, a_i_loc + 1, b_i_loc + 1, len, retdest
+    EQ
+    ISZERO
+    %jumpi(add_loop)
+add_end:
+    // stack: carry_new, i + 1, a_i_loc + 1, b_i_loc + 1, len, retdest
+    %stack (c, i, a, b, n) -> (c, a)
+    // stack: carry_new, a_i_loc + 1, retdest
+    // If carry = 0, no need to increment.
+    ISZERO
+    %jumpi(increment_end)
+increment_loop:
+    // stack: cur_loc, retdest
+    DUP1
+    %mload_kernel_general
+    // stack: val, cur_loc, retdest
+    %increment
+    // stack: val+1, cur_loc, retdest
+    SWAP1
+    DUP2
+    DUP2
+    // stack: cur_loc, val+1, cur_loc, val+1, retdest
+    %mstore_kernel_general
+    // stack: cur_loc, val+1, retdest
+    %increment
+    // stack: cur_loc + 1, val+1, retdest
+    SWAP1
+    // stack: val+1, cur_loc + 1, retdest
+    %eq_const(@BIGNUM_LIMB_BASE)
+    ISZERO
+    %jumpi(increment_end)
+    // stack: cur_loc + 1, retdest
+    PUSH 0
+    DUP2
+    // stack: cur_loc + 1, 0, cur_loc + 1, retdest
+    %mstore_kernel_general
+    %jump(increment_loop)
```

Comment:
Ah, well the `%eq_const(@BIGNUM_LIMB_BASE)` is what checks for overflow, and in that case it replaces it with zero and then keeps doing the `increment` loop (that is, keeps on carrying). So in theory it can keep on looping, but only if each following limb has the maximum value of 2^128 - 1

I think it's correct, but I might have missed something; but either way, your idea of changing the interface sounds good! I'll discuss with Daniel

Reply:
Hmm, consider the case `a` and `b` have length 1 and `a + b >= 2^128` so the carry is 1. It looks to me that on lines 58--60 `cur_loc = a + 1` which is already an invalid address, right?

Even if it isn't an invalid address, in general the value `val` read on line 60 could be anything up to `2^128 - 1`; if it's the maximum, then the `%increment` on line 62 causes `val` to be 2^128 and this is then stored at `cur_loc` on line 68. But *limbs must be strictly less than 2^128 to be properly normalised.*

This also means the test on line 74 is not correct; it's true that `val + 1 == BIGNUM_LIMB_BASE` indicates overflow, but the overflowing value `val + 1` has already been stored and `cur_loc` has been incremented to point to the next limb.

So I think this code creates invalid (or at least not properly normalised) bignums in the case where the addition overflows. I think it would be much easier to delete the whole `increment_loop` section and just return the carry, but if you want to stick to this approach you'll need to (i) ensure that `a` has space `len + 1` so there's space for the carry (this already creates the problem of different length inputs) and (ii) the increment loop isn't necessary anyway since if `a` and `b` start with the same length then they can at most overflow by a single bit, so the loop should never run more than one iteration (though this isn't true if the inputs have different lengths).

Hope this is clear; let me know if I've misunderstood anything or made a mistake somewhere! :)

Reply:
Ohh I see, thanks! Ah the point of storing the 0 was to overwrite the overflowed 2^128 value, and then increment the next one; but I incorrectly store the 0 in `cur_loc + 1` instead of `cur_loc`. But you're right that the loop isn't necessary; I'll change it to return the carry bit!